### ISSUE TITLE
Add climographs to the climatic menu

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#8657]
@jkmusyoka  Here is the Climatic > Describe menu.  It is very poor.

<img width="306" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/1d22ee67-4e74-43a5-9c35-500d58db1d5e">

If you ask for climatic graphs it suggests bar charts and climographs are popular.  The recent dodoma summaries provide data with just monthly summaries of rainfall and temperatures.  I was not able to get a climaograph from the Describe > Graphs > General dialog - perhaps you can?

In addition could we have various attractive climographs perhaps using R code you suggest, but I would like to see what chatGPT could come up with?  I assume that will be built on code and hope it could suggest several different alternatives?

Then could we quickly construct an impressive dialog that implements these alternatives.

The basic structure is clear.

a) Usual data selector
b) Single receiver with label Rainfall.
c) Multiple receiver with label Temperature(s)  (It could just have mean monthly temperatures or Mean min and mean max, and perhaps other possibilities including extremes as well.
d) I assume there might be a second axis label to allow Multiplier and Offset for the secondary axis?
e) Be able to save the graph of course.
f) Other options dependin g on what there are.  For example maybe the max and min could have bands? Could there be an Alphs control to help the visibility of the different components?

I am keen of checking the sort of AI process as well as the dialog.

And I wonder about a Walter-Leith graph, from climatol, on another button on the same tab.  I think it can use the same data?

I would also like to buuld that Climatic > Describe menu more.

**Comment by rdstern** [comment]
@jkmusyoka I have now found a ggplot2 version of the Walter-Leith plot.  It is in the climaetol package.  However that needs an API to be added and still has the data in a same odd layout (months going across).  We would want to adapt it to be able to handle multiple stations!

I propose it may be simpler to adapt their function to cope with our layout of the data, namely months going down, and possibly also multiple stations - which are the obvious facets!

It is the `ggclimat_walter_lieth` function.  Here is their code in case that helps!

```
function (dat, est = "", alt = NA, per = NA, mlab = "es", pcol = "#002F70", 
    tcol = "#ff0000", pfcol = "#9BAEE2", sfcol = "#3C6FC4", shem = FALSE, 
    p3line = FALSE, ...) 
{
    if (!all(dim(dat) == c(4, 12))) {
        stop("`dat` should have 4 rows and 12 colums. Your inputs has ", 
            nrow(dat), " rows and ", ncol(dat), " columns.")
    }
    data_na <- as.integer(sum(is.na(dat)))
    if (data_na > 0) {
        stop("Data with null values, unable to plot the diagram \n")
    }
    if (is.matrix(dat)) {
        dat <- as.data.frame(dat, row.names = c("p_mes_md", "tm_max_md", 
            "tm_min_md", "ta_min_min"), col.names = paste0("m", 
            seq_len(12)))
    }
    mlab <- toupper(substr(readr::locale(mlab)$date_names$mon, 
        1, 1))
    dat_long <- tibble::as_tibble(as.data.frame(t(dat)))
    names(dat_long) <- c("p_mes", "tm_max", "tm_min", "ta_min")
    dat_long <- dplyr::bind_cols(label = mlab, dat_long)
    if (shem) {
        dat_long <- rbind(dat_long[7:12, ], dat_long[1:6, ])
    }
    dat_long$tm <- (dat_long[[3]] + dat_long[[4]])/2
    dat_long$pm_reesc <- ifelse(dat_long$p_mes < 100, dat_long$p_mes * 
        0.5, dat_long$p_mes * 0.05 + 45)
    dat_long$p3line <- dat_long$p_mes/3
    dat_long <- dplyr::bind_rows(dat_long[nrow(dat_long), ], 
        dat_long, dat_long[1, ])
    dat_long[c(1, nrow(dat_long)), "label"] <- ""
    dat_long <- cbind(indrow = seq(-0.5, 12.5, 1), dat_long)
    dat_long_int <- NULL
    for (j in seq(nrow(dat_long) - 1)) {
        intres <- NULL
        for (i in seq_len(ncol(dat_long))) {
            if (is.character(dat_long[j, i])) {
                val <- as.data.frame(dat_long[j, i])
            }
            else {
                interpol <- approx(x = dat_long[c(j, j + 1), 
                  1], y = dat_long[c(j, j + 1), i], n = 50)
                val <- as.data.frame(interpol$y)
            }
            names(val) <- names(dat_long)[i]
            intres <- dplyr::bind_cols(intres, val)
        }
        dat_long_int <- dplyr::bind_rows(dat_long_int, intres)
    }
    dat_long_int$interpolate <- TRUE
    dat_long_int$label <- ""
    dat_long$interpolate <- FALSE
    dat_long_int <- dat_long_int[!dat_long_int$indrow %in% dat_long$indrow, 
        ]
    dat_long_end <- dplyr::bind_rows(dat_long, dat_long_int)
    dat_long_end <- dat_long_end[order(dat_long_end$indrow), 
        ]
    dat_long_end <- dat_long_end[dat_long_end$indrow >= 0 & dat_long_end$indrow <= 
        12, ]
    dat_long_end <- tibble::as_tibble(dat_long_end)
    month_breaks <- dat_long_end[dat_long_end$label != "", ]$indrow
    month_labs <- dat_long_end[dat_long_end$label != "", ]$label
    ymax <- max(60, 10 * floor(max(dat_long_end$pm_reesc)/10) + 
        10)
    ymin <- min(-3, min(dat_long_end$tm))
    range_tm <- seq(0, ymax, 10)
    if (ymin < -3) {
        ymin <- floor(ymin/10) * 10
        range_tm <- seq(ymin, ymax, 10)
    }
    templabs <- paste0(range_tm)
    templabs[range_tm > 50] <- ""
    range_prec <- range_tm * 2
    range_prec[range_tm > 50] <- range_tm[range_tm > 50] * 20 - 
        900
    preclabs <- paste0(range_prec)
    preclabs[range_tm < 0] <- ""
    title <- est
    if (!is.na(alt)) {
        title <- paste0(title, " (", prettyNum(alt, big.mark = ",", 
            decimal.mark = "."), " m)")
    }
    if (!is.na(per)) {
        title <- paste0(title, "\n", per)
    }
    sub <- paste(round(mean(dat_long_end[dat_long_end$interpolate == 
        FALSE, ]$tm), 1), "C        ", prettyNum(round(sum(dat_long_end[dat_long_end$interpolate == 
        FALSE, ]$p_mes)), big.mark = ","), " mm", sep = "")
    maxtm <- prettyNum(round(max(dat_long_end$tm_max), 1))
    mintm <- prettyNum(round(min(dat_long_end$tm_min), 1))
    tags <- paste0(paste0(rep(" \n", 6), collapse = ""), maxtm, 
        paste0(rep(" \n", 10), collapse = ""), mintm)
    ticks <- data.frame(x = seq(0, 12), ymin = -3, ymax = 0)
    getpolymax <- function(x, y, y_lim) {
        initpoly <- FALSE
        yres <- NULL
        xres <- NULL
        for (i in seq_len(length(y))) {
            lastobs <- i == length(x)
            if (y[i] > y_lim[i]) {
                if (isFALSE(initpoly)) {
                  xres <- c(xres, x[i])
                  yres <- c(yres, y_lim[i])
                  initpoly <- TRUE
                }
                xres <- c(xres, x[i])
                yres <- c(yres, y[i])
                if (lastobs) {
                  xres <- c(xres, x[i], NA)
                  yres <- c(yres, y_lim[i], NA)
                }
            }
            else {
                if (initpoly) {
                  xres <- c(xres, x[i - 1], NA)
                  yres <- c(yres, y_lim[i - 1], NA)
                  initpoly <- FALSE
                }
            }
        }
        poly <- tibble::tibble(x = xres, y = yres)
        return(poly)
    }
    getlines <- function(x, y, y_lim) {
        yres <- NULL
        xres <- NULL
        ylim_res <- NULL
        for (i in seq_len(length(y))) {
            if (y[i] > y_lim[i]) {
                xres <- c(xres, x[i])
                yres <- c(yres, y[i])
                ylim_res <- c(ylim_res, y_lim[i])
            }
        }
        line <- tibble::tibble(x = xres, y = yres, ylim_res = ylim_res)
        return(line)
    }
    prep_max_poly <- getpolymax(x = dat_long_end$indrow, y = pmax(dat_long_end$pm_reesc, 
        50), y_lim = rep(50, length(dat_long_end$indrow)))
    tm_max_line <- getlines(x = dat_long_end$indrow, y = dat_long_end$tm, 
        y_lim = dat_long_end$pm_reesc)
    pm_max_line <- getlines(x = dat_long_end$indrow, y = pmin(dat_long_end$pm_reesc, 
        50), y_lim = dat_long_end$tm)
    dat_real <- dat_long_end[dat_long_end$interpolate == FALSE, 
        c("indrow", "ta_min")]
    x <- NULL
    y <- NULL
    for (i in seq_len(nrow(dat_real))) {
        if (dat_real[i, ]$ta_min < 0) {
            x <- c(x, NA, rep(dat_real[i, ]$indrow - 0.5, 2), 
                rep(dat_real[i, ]$indrow + 0.5, 2), NA)
            y <- c(y, NA, -3, 0, 0, -3, NA)
        }
        else {
            x <- c(x, NA)
            y <- c(y, NA)
        }
    }
    probfreeze <- tibble::tibble(x = x, y = y)
    rm(dat_real)
    dat_real <- dat_long_end[dat_long_end$interpolate == FALSE, 
        c("indrow", "tm_min")]
    x <- NULL
    y <- NULL
    for (i in seq_len(nrow(dat_real))) {
        if (dat_real[i, ]$tm_min < 0) {
            x <- c(x, NA, rep(dat_real[i, ]$indrow - 0.5, 2), 
                rep(dat_real[i, ]$indrow + 0.5, 2), NA)
            y <- c(y, NA, -3, 0, 0, -3, NA)
        }
        else {
            x <- c(x, NA)
            y <- c(y, NA)
        }
    }
    surefreeze <- tibble::tibble(x = x, y = y)
    wandlplot <- ggplot2::ggplot() + ggplot2::geom_line(data = dat_long_end, 
        aes(x = .data$indrow, y = .data$pm_reesc), color = pcol) + 
        ggplot2::geom_line(data = dat_long_end, aes(x = .data$indrow, 
            y = .data$tm), color = tcol)
    if (nrow(tm_max_line > 0)) {
        wandlplot <- wandlplot + ggplot2::geom_segment(aes(x = .data$x, 
            y = .data$ylim_res, xend = .data$x, yend = .data$y), 
            data = tm_max_line, color = tcol, alpha = 0.2)
    }
    if (nrow(pm_max_line > 0)) {
        wandlplot <- wandlplot + ggplot2::geom_segment(aes(x = .data$x, 
            y = .data$ylim_res, xend = .data$x, yend = .data$y), 
            data = pm_max_line, color = pcol, alpha = 0.2)
    }
    if (p3line) {
        wandlplot <- wandlplot + ggplot2::geom_line(data = dat_long_end, 
            aes(x = .data$indrow, y = .data$p3line), color = pcol)
    }
    if (max(dat_long_end$pm_reesc) > 50) {
        wandlplot <- wandlplot + ggplot2::geom_polygon(data = prep_max_poly, 
            aes(x, y), fill = pcol)
    }
    if (min(dat_long_end$ta_min) < 0) {
        wandlplot <- wandlplot + ggplot2::geom_polygon(data = probfreeze, 
            aes(x = x, y = y), fill = pfcol, colour = "black")
    }
    if (min(dat_long_end$tm_min) < 0) {
        wandlplot <- wandlplot + geom_polygon(data = surefreeze, 
            aes(x = x, y = y), fill = sfcol, colour = "black")
    }
    wandlplot <- wandlplot + geom_hline(yintercept = c(0, 50), 
        size = 0.5) + geom_segment(data = ticks, aes(x = x, xend = x, 
        y = ymin, yend = ymax)) + scale_x_continuous(breaks = month_breaks, 
        name = "", labels = month_labs, expand = c(0, 0)) + scale_y_continuous("C", 
        limits = c(ymin, ymax), labels = templabs, breaks = range_tm, 
        sec.axis = dup_axis(name = "mm", labels = preclabs))
    wandlplot <- wandlplot + ggplot2::labs(title = title, subtitle = sub, 
        tag = tags) + ggplot2::theme_classic() + ggplot2::theme(plot.title = element_text(lineheight = 1, 
        size = 14, face = "bold"), plot.subtitle = element_text(hjust = 1, 
        vjust = 1, size = 14), plot.tag = element_text(size = 10), 
        plot.tag.position = "left", axis.ticks.length.x.bottom = unit(0, 
            "pt"), axis.line.x.bottom = element_blank(), axis.title.y.left = element_text(angle = 0, 
            vjust = 0.9, size = 10, colour = tcol, margin = unit(rep(10, 
                4), "pt")), axis.text.x.bottom = element_text(size = 10), 
        axis.text.y.left = element_text(colour = tcol, size = 10), 
        axis.title.y.right = element_text(angle = 0, vjust = 0.9, 
            size = 10, colour = pcol, margin = unit(rep(10, 4), 
                "pt")), axis.text.y.right = element_text(colour = pcol, 
            size = 10))
    return(wandlplot)
}
<bytecode: 0x00000299dfb674a0>
<environment: namespace:climaemet>
```

**Comment by Vitalis95** [comment]
@jkmusyoka , What do you think about this?

**Comment by rdstern** [comment]
@lilyclements I am now continuing with this issue just for the Walter-Leith option of this dialog.  It hasn't got far as you have seen.  I still assume 2 radio buttons on the (I assume empty) Climograph dialog.  This is the second, and I suggest we start with it.
I think you are right, that the function above does transform the odd layout of the data into the shape we start with.  So I hope the change in Function will be easy.  I'm not sure about the adaptation to multiple stations.  

I think the dialog may be similar to the Seasonal one, so you may wish to work with @MeSophie, so she could adapt that one.
There may be 3 single receivers for the y-variables, though perhaps a multiple receiver will be simpler,  and I think they may fill automatically, for data defined as climatic.  The x, by default could be the month variable.  Then the control for the facets could be filled with the station - as for the seasonal plot.

**Comment by lilyclements** [comment]
@rdstern I've been playing around with this code tonight. Their code can do with some work, but I assume that's not a priority. I've played around a bit and we can now use it for multiple stations -- at least for the data they gave!

There's two functions - the first `prepare_walter_leith` is called in the ggplot function so you don't need to worry about that. This one just manipulates different data sets needed for our ggplot.

```
prepare_walter_lieth <- function(data, month, tm_min, ta_min){
  dat_long_int <- NULL
  for (j in seq(nrow(data) - 1)) {
    intres <- NULL
    for (i in seq_len(ncol(data))) {
      if (is.character(data[j, i])) {
        val <- as.data.frame(data[j, i])
      }
      else {
        interpol <- approx(x = data[c(j, j + 1), "indrow"],
                           y = data[c(j, j + 1), i],
                           n = 50)
        val <- as.data.frame(interpol$y)
      }
      names(val) <- names(data)[i]
      intres <- dplyr::bind_cols(intres, val)
    }
    dat_long_int <- dplyr::bind_rows(dat_long_int, intres)
  }
  dat_long_int$interpolate <- TRUE
  dat_long_int[[month]] <- ""
  data$interpolate <- FALSE
  dat_long_int <- dat_long_int[!dat_long_int$indrow %in% data$indrow, ]
  dat_long_end <- dplyr::bind_rows(data, dat_long_int)
  dat_long_end <- dat_long_end[order(dat_long_end$indrow), ]
  dat_long_end <- dat_long_end[dat_long_end$indrow >= 0 & dat_long_end$indrow <= 12, ]
  dat_long_end <- tibble::as_tibble(dat_long_end)
  
  getpolymax <- function(x, y, y_lim) {
    initpoly <- FALSE
    yres <- NULL
    xres <- NULL
    for (i in seq_len(length(y))) {
      lastobs <- i == length(x)
      if (y[i] > y_lim[i]) {
        if (isFALSE(initpoly)) {
          xres <- c(xres, x[i])
          yres <- c(yres, y_lim[i])
          initpoly <- TRUE
        }
        xres <- c(xres, x[i])
        yres <- c(yres, y[i])
        if (lastobs) {
          xres <- c(xres, x[i], NA)
          yres <- c(yres, y_lim[i], NA)
        }
      }
      else {
        if (initpoly) {
          xres <- c(xres, x[i - 1], NA)
          yres <- c(yres, y_lim[i - 1], NA)
          initpoly <- FALSE
        }
      }
    }
    poly <- tibble::tibble(x = xres, y = yres)
    return(poly)
  }
  getlines <- function(x, y, y_lim) {
    yres <- NULL
    xres <- NULL
    ylim_res <- NULL
    for (i in seq_len(length(y))) {
      if (y[i] > y_lim[i]) {
        xres <- c(xres, x[i])
        yres <- c(yres, y[i])
        ylim_res <- c(ylim_res, y_lim[i])
      }
    }
    line <- tibble::tibble(x = xres, y = yres, ylim_res = ylim_res)
    return(line)
  }
  prep_max_poly <- getpolymax(x = dat_long_end$indrow, y = pmax(dat_long_end$pm_reesc, 
                                                                50), y_lim = rep(50, length(dat_long_end$indrow)))
  tm_max_line <- getlines(x = dat_long_end$indrow, y = dat_long_end$tm, 
                          y_lim = dat_long_end$pm_reesc)
  pm_max_line <- getlines(x = dat_long_end$indrow, y = pmin(dat_long_end$pm_reesc, 
                                                            50), y_lim = dat_long_end$tm)
  dat_real <- dat_long_end[dat_long_end$interpolate == FALSE, 
                           c("indrow", ta_min)]
  x <- NULL
  y <- NULL
  for (i in seq_len(nrow(dat_real))) {
    if (dat_real[i, ][[ta_min]] < 0) {
      x <- c(x, NA, rep(dat_real[i, ]$indrow - 0.5, 2), 
             rep(dat_real[i, ]$indrow + 0.5, 2), NA)
      y <- c(y, NA, -3, 0, 0, -3, NA)
    }
    else {
      x <- c(x, NA)
      y <- c(y, NA)
    }
  }
  probfreeze <- tibble::tibble(x = x, y = y)
  rm(dat_real)
  dat_real <- dat_long_end[dat_long_end$interpolate == FALSE, 
                           c("indrow", tm_min)]
  x <- NULL
  y <- NULL
  for (i in seq_len(nrow(dat_real))) {
    if (dat_real[i, ][[tm_min]] < 0) {
      x <- c(x, NA, rep(dat_real[i, ]$indrow - 0.5, 2), 
             rep(dat_real[i, ]$indrow + 0.5, 2), NA)
      y <- c(y, NA, -3, 0, 0, -3, NA)
    }
    else {
      x <- c(x, NA)
      y <- c(y, NA)
    }
  }
  surefreeze <- tibble::tibble(x = x, y = y)
  return_list <- list(dat_long_end,
                      tm_max_line,
                      pm_max_line,
                      prep_max_poly,
                      probfreeze,
                      surefreeze)
  names(return_list) <- c("dat_long_end", "tm_max_line", "pm_max_line",
                          "prep_max_poly", "prob_freeze", "surefreeze")
  return(return_list)
}

ggwalter_lieth <- function (data, month, station, p_mes, tm_max, tm_min, ta_min, station_name = "", 
                            alt = NA, per = NA, pcol = "#002F70", 
                            tcol = "#ff0000", pfcol = "#9BAEE2", sfcol = "#3C6FC4", 
                            shem = FALSE, p3line = FALSE, ...) 
{
  
  # Preprocess data with vectorised operations
  data <- data %>%
    dplyr::mutate(tm = (.data[[tm_max]] + .data[[tm_min]]) / 2,
                  pm_reesc = dplyr::if_else(.data[[p_mes]] < 100, .data[[p_mes]] * 0.5, .data[[p_mes]] * 0.05 + 45),
                  p3line = .data[[p_mes]] / 3) %>%
    dplyr::mutate(across(.data[[month]], ~ fct_expand(.data[[month]], ""))) %>%
    dplyr::arrange(.data[[month]])
  
  # do this for each station, if we have a station
  if (!is.null(station)){
    data <- data %>% group_by(station)
  }
  data <- data %>%
    group_modify(~{
      # Add dummy rows at the beginning and end for each group
      .x <- bind_rows(.x[nrow(.x), , drop = FALSE], .x, .x[1, , drop = FALSE])
      # Clear month value for the dummy rows
      .x[c(1, nrow(.x)), which(names(.x) == data[[month]])] <- ""
      # Add an index column for plotting or further transformations
      .x <- cbind(indrow = seq(-0.5, 12.5, 1), .x)
      .x
    })
  
  if (!is.null(station)){
    data <- data %>% ungroup()
  }
  data <- data.frame(data)
  
  # split by station
  if (is.null(station)){
    data_list <- prepare_walter_lieth(data, month, tm_min, ta_min)
    # data things
    dat_long_end <- data_list$dat_long_end
    tm_max_line <- data_list$tm_max_line
    pm_max_line <- data_list$pm_max_line
    prep_max_poly <- data_list$prep_max_poly
    probfreeze <- data_list$prob_freeze
    surefreeze <- data_list$surefreeze
  } else {
    results <-
      map(.x = unique(data[[station]]),
          .f = ~{filtered_data <- data %>% filter(!!sym(station) == .x)
          prepare_walter_lieth(filtered_data, month, tm_min, ta_min)})
    # Function to bind rows for a specific sub-element across all main elements
    n <- length(results)
    m <- length(results[[1]])
    station_name <- unique(data[[station]])
    binds <- NULL
    combined <- NULL
    for (j in 1:m){
      for (i in 1:n) { # for each station data set
        binds[[i]] <- results[[i]][[j]] %>% mutate(!!sym(station) := station_name[i])
      }
      combined[[j]] <- do.call(rbind, binds) # Combine all the sub-elements row-wise
    }
    # data things
    dat_long_end <- combined[[1]]
    tm_max_line <- combined[[2]]
    pm_max_line <- combined[[3]]
    prep_max_poly <- combined[[4]]
    probfreeze <- combined[[5]]
    surefreeze <- combined[[6]]
  }
  
  # if (is.null(station)){
  #   run fun
  # } else {
  #   for (i in station){
  #   print(i)
  #   data <- dat %>% filter(station == i) # todo: set station to be a variable not hard coded 
  #   run fun}
  # }
  
  # data frame pretty things ------------------------------------------------------
  ticks <- data.frame(x = seq(0, 12), ymin = -3, ymax = 0)
  title <- station_name
  if (!is.na(alt)) {
    title <- paste0(title, " (", prettyNum(alt, big.mark = ",", 
                                           decimal.mark = "."), " m)")
  }
  if (!is.na(per)) {
    title <- paste0(title, "\n", per)
  }
  sub <- paste(round(mean(dat_long_end[dat_long_end$interpolate == FALSE, ]$tm), 1),
               "C        ",
               prettyNum(round(sum(dat_long_end[dat_long_end$interpolate == FALSE, ][[p_mes]])), big.mark = ","), " mm", sep = "")
  
  maxtm <- prettyNum(round(max(dat_long_end[[tm_max]]), 1))
  mintm <- prettyNum(round(min(dat_long_end[[tm_min]]), 1))
  tags <- paste0(paste0(rep(" \n", 6), collapse = ""), maxtm, 
                 paste0(rep(" \n", 10), collapse = ""), mintm)
  month_breaks <- dat_long_end[dat_long_end[[month]] != "", ]$indrow
  month_labs <- dat_long_end[dat_long_end[[month]] != "", ][[month]]
  
  ymax <- max(60, 10 * floor(max(dat_long_end$pm_reesc)/10) + 10)
  ymin <- min(-3, min(dat_long_end$tm))
  range_tm <- seq(0, ymax, 10)
  if (ymin < -3) {
    ymin <- floor(ymin/10) * 10
    range_tm <- seq(ymin, ymax, 10)
  }
  templabs <- paste0(range_tm)
  templabs[range_tm > 50] <- ""
  range_prec <- range_tm * 2
  range_prec[range_tm > 50] <- range_tm[range_tm > 50] * 20 - 900
  preclabs <- paste0(range_prec)
  preclabs[range_tm < 0] <- ""
  
  wandlplot <- ggplot2::ggplot() + ggplot2::geom_line(data = dat_long_end, 
                                                      aes(x = .data$indrow, y = .data$pm_reesc), color = pcol) + 
    ggplot2::geom_line(data = dat_long_end, aes(x = .data$indrow, 
                                                y = .data$tm), color = tcol)
  if (nrow(tm_max_line > 0)) {
    wandlplot <- wandlplot + ggplot2::geom_segment(aes(x = .data$x, 
                                                       y = .data$ylim_res, xend = .data$x, yend = .data$y), 
                                                   data = tm_max_line, color = tcol, alpha = 0.2)
  }
  if (nrow(pm_max_line > 0)) {
    wandlplot <- wandlplot + ggplot2::geom_segment(aes(x = .data$x, 
                                                       y = .data$ylim_res, xend = .data$x, yend = .data$y), 
                                                   data = pm_max_line, color = pcol, alpha = 0.2)
  }
  if (p3line) {
    wandlplot <- wandlplot + ggplot2::geom_line(data = dat_long_end, 
                                                aes(x = .data$indrow, y = .data$p3line), color = pcol)
  }
  if (max(dat_long_end$pm_reesc) > 50) {
    wandlplot <- wandlplot + ggplot2::geom_polygon(data = prep_max_poly, aes(x, y),
                                                   fill = pcol)
  }
  if (min(dat_long_end[[ta_min]]) < 0) {
    wandlplot <- wandlplot + ggplot2::geom_polygon(data = probfreeze, aes(x = x, y = y),
                                                   fill = pfcol, colour = "black")
  }
  if (min(dat_long_end[[tm_min]]) < 0) {
    wandlplot <- wandlplot + geom_polygon(data = surefreeze, aes(x = x, y = y),
                                          fill = sfcol, colour = "black")
  }
  wandlplot <- wandlplot + geom_hline(yintercept = c(0, 50), 
                                      size = 0.5) +
    geom_segment(data = ticks, aes(x = x, xend = x, y = ymin, yend = ymax)) +
    scale_x_continuous(breaks = month_breaks, name = "", labels = month_labs, expand = c(0, 0)) + 
    scale_y_continuous("C", limits = c(ymin, ymax), labels = templabs, 
                       breaks = range_tm, sec.axis = dup_axis(name = "mm", labels = preclabs))
  wandlplot <- wandlplot +
    ggplot2::labs(title = title, subtitle = sub, tag = tags) +
    ggplot2::theme_classic() +
    ggplot2::theme(plot.title = element_text(lineheight = 1, size = 14, face = "bold"),
                   plot.subtitle = element_text(hjust = 1, vjust = 1, size = 14),
                   plot.tag = element_text(size = 10), 
                   plot.tag.position = "left", axis.ticks.length.x.bottom = unit(0, "pt"), 
                   axis.line.x.bottom = element_blank(), 
                   axis.title.y.left = element_text(angle = 0, 
                                                    vjust = 0.9, size = 10, colour = tcol,
                                                    margin = unit(rep(10, 4), "pt")),
                   axis.text.x.bottom = element_text(size = 10), 
                   axis.text.y.left = element_text(colour = tcol, size = 10), 
                   axis.title.y.right = element_text(angle = 0, vjust = 0.9, 
                                                     size = 10, colour = pcol,
                                                     margin = unit(rep(10, 4), "pt")),
                   axis.text.y.right = element_text(colour = pcol, size = 10))
  
  if (!is.null(station)){
    wandlplot <- wandlplot + facet_wrap(station)
  }
  
  return(wandlplot)
}
```

Sorry about the length there!

So to test it:

```
# get the data
dat <- climaemet::climaemet_9434_climatogram

# put it into the format we will have
dat_long <- tibble::as_tibble(as.data.frame(t(dat))) 

# Add a month variable
dat_long$month <- forcats::as_factor(c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))

# now I'll just double it, and make station A and B to illustrate:
dat_long1 <- rbind(dat_long, dat_long)
dat_long1$station <- c(rep("A", 12), rep("B", 12))

# then to run it
ggwalter_lieth(data = dat_long1, month = "month", station = "station",
               "p_mes_md", "tm_max_md", "tm_min_md", "ta_min_min")
```

1. I couldn't test with other data because I'm not sure what `ta_min` is. What is `ta_min`?
2. Are "p_mes", "tm_max" and "tm_min" - rain, max temp and min temp?
3. Their function offers a variable to retitle the plot. I assume we don't want that - or we can give our own hard-coded title, which the user can change in a sub-dialog (similarly for subtitle)

**Comment by rdstern** [comment]
@lilyclements great - just great.  I am looking now.

**Comment by MeSophie** [comment]
@lilyclements  Please I have some questions :
a) I assume the long code you sent is for stand alone function and the short one the way the code should be implemented in R-Instat. Am I right?

b) Base on this [document](https://www.rdocumentation.org/packages/climaemet/versions/1.0.1/topics/climaemet_9434_climatogram), as you said `"p_mes", "tm_max", "tm_min",  and  ta_min`  are rain, max temp, min temp and Absolute monthly minimum temperature. Now my question is do you need a multiple receiver for these parameters as @rdstern mentioned earlier, if yes which parameter this multiple receiver will pass?

**Comment by lilyclements** [comment]
@MeSophie yes, we should put this in the `stand_alone_functions.R` file. 

Thank you for finding that document and clarifying those parameters - I was confused about `ta_min`! In here, we would currently need six single receivers:

* month
* station
* p_mes
* tm_max
* tm_min
* ta_min - it could be that we calculate this in the function?

@rdstern How is `ta_min` calculated? Can we calculate this measure from `tm_min`?

**Comment by MeSophie** [comment]
@lilyclements Thank you.

**Comment by MeSophie** [comment]
@lilyclements I hope you can help. Here is the current status of the dialog.

![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/73c39a24-f401-4343-a9a3-d727a38b6f17)

When I click on to script this is the code that it produces:
```
# Dialog: Climograph

dodoma <- data_book$get_data_frame(data_name="dodoma")
dat_long <- tibble::as_tibble(as.data.frame(t(dodoma)))

dat_long$month <- forcats::as_factor(c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))

dat_long1 <- rbind(dat_long, dat_long)

dat_long1$station <- c(rep("A", 12), rep("B", 12))

last_graph <- ggwalter_lieth(data=dat_long1, month="month_abbr", p_mes="rain", tm_max="tmax", tm_min="tmin")
data_book$add_object(data_name="dodoma", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))

data_book$get_object_data(data_name="dodoma", as_file=TRUE)
rm(list=c("dat_long", "dodoma", "dat_long$month", "dat_long1", "dat_long1$station", "last_graph"))
```
The first time I directly click on OK, I had this error message:

![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/8f353444-3c33-481d-a352-eb353cf452b9)
![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/b78362c1-294f-42ca-87ac-b2eed56216b4)
![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/faa21b08-633d-4d03-a6b7-69bb1d3928cf)
 
So I went to script windows to investigate if the problem is dodoma data. I import climaemet_9434_climatogram data then edit the code 
```
climaemet_9434_climatogram <- data_book$get_data_frame(data_name="climaemet_9434_climatogram")
dat_long <- tibble::as_tibble(as.data.frame(t(climaemet_9434_climatogram)))
dat_long$month <- forcats::as_factor(c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))
dat_long1 <- rbind(dat_long, dat_long)
dat_long1$station <- c(rep("A", 12), rep("B", 12))
last_graph <- ggwalter_lieth(data = dat_long1, month = "month", station = "station","p_mes_md", "tm_max_md", "tm_min_md", "ta_min_min")
data_book$add_object(data_name="climaemet_9434_climatogram", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="climaemet_9434_climatogram", as_file=TRUE)
rm(list=c("dat_long", "climaemet_9434_climatogram", "dat_long$month", "dat_long1", "dat_long1$station", "last_graph"))
```



![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/a612d263-a868-4a80-b0ae-164f222311bf)

**Comment by rdstern** [comment]
Hi @MeSophie I get the same error as above, but I think we can proceed a bit, befor (or in parallel) with bringing @lilyclements back into the discussion.

I attach a data file you can use.

**Comment by lilyclements** [comment]
@MeSophie you're actually doing _too much_ here. 

I gave this code before:

```
climaemet_9434_climatogram <- data_book$get_data_frame(data_name="climaemet_9434_climatogram")
dat_long <- tibble::as_tibble(as.data.frame(t(climaemet_9434_climatogram)))
dat_long$month <- forcats::as_factor(c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"))
dat_long1 <- rbind(dat_long, dat_long)
dat_long1$station <- c(rep("A", 12), rep("B", 12))
last_graph <- ggwalter_lieth(data = dat_long1, month = "month", station = "station","p_mes_md", "tm_max_md", "tm_min_md", "ta_min_min")
data_book$add_object(data_name="climaemet_9434_climatogram", object_name="last_graph", object_type_label="graph", object_format="image", object=check_graph(graph_object=last_graph))
data_book$get_object_data(data_name="climaemet_9434_climatogram", as_file=TRUE)
rm(list=c("dat_long", "climaemet_9434_climatogram", "dat_long$month", "dat_long1", "dat_long1$station", "last_graph"))
```

Anything involving `dat_long` (so lines 2-5) can be completely ignored since this is just unique to the `climaemet_9434_climatogram` data.
* On line 2, I'm transposing our climaemet_9434_climatogram data frame. However, with dodoma we don't need to do this.
* On line 3, I'm setting `month` to be a factor, but again, we don't need to do this with dodoma.
* On line 4, I'm just duplicating our data so that I can give an example with stations
* On line 5, I'm just adding a dummy variable with station as "A" for the first 12 lines, and "B" for the second 12 lines.

I'm sorry I didn't make that clearer before. This is all stuff that the user can do in other dialogs in R-Instat when they get their data in that format. 

For your `dodoma_by_month_abbr` data, you're already in the right format - so you've done it! This can just go into `ggwalter_leith`

```
last_graph <- ggwalter_lieth(data=dodoma_by_month_abbr, month="month_abbr",
                             p_mes="mean_sum_rain",
                             tm_max="mean_tmax",
                             tm_min="mean_tmin",
                             ta_min = "mean_tmin")
```

So, the only R code that needs to be run here is the `ggwalter_lieth` code.

All the other stuff is just me getting `climaemet_9434_climatogram` into the right format. The format that climaemet_9434_climatogram is in is unusual to use in statistical software.
So we can assume that the user can do this themselves using other dialogs in R-Instat, as most times they should have the data in the right format if they've been using R.

**Comment by lilyclements** [comment]
To add - @MeSophie you are absolutely correct with the last error you gave.

`fct_expand` is from the `forcats` package. Can you amend the function to say `forcats::fct_expand` where it currently says `fct_expand`?
